Fix profile status message overflow & subscription

rarira opened the pull request:
## Specify project
Both client and server side

## Description

There have been a issue[#455] which profile status message shown in both `MainFriend` and `ProfileModal`. And I think it might be better if my friend updates his/her status message in `ProfileUpdate`, the status message which I'm seeing also should be updated in real-time.

So I changed and added some codes

1. Set the limit of statusMessage Input of `ProfileUpdate`
    ````
          <EditText
            testID="input-status"
            type="column"
            style={{marginTop: 36}}

            .......

            textInputProps={{
              multiline: true,
              maxLength: 60,
            }}
    ````
2. Add `numberOfLines` and `ellipsizeMode` attribute to `UserListItem`
   ```
          ) : statusMessage ? (
            <StyledRightText numberOfLines={1} ellipsizeMode={'tail'}>
              {statusMessage}
            </StyledRightText>

   ```
3.  Added new UI component `StatusMessageView` to display animation of statusMessage in `ProfileModal` 
![Simulator Screen Recording - iPhone 12 Pro - 2021-08-17 at 22 21 56](https://user-images.githubusercontent.com/8540420/129733304-ad5d5102-c63a-4b8f-94b9-275589d30a43.gif)

I added arrow icon if there _might_ be something more statusMessage to show, but because the logic is quite simple, there can be unexpected cases and so even if there is nothing more, the animation still works without arrow. I researched a lot about this issue, there seems to be no perfect solution on this issue, applied to all platforms(web, ios, android)
https://github.com/necolas/react-native-web/issues/814

4. Added `useSubscription` to `MainFriend` 

## Related Issues

[#455] 

## Tests

I tried to write tests code of newly added `StatusMessageView` component for myself this time.
But I'm really sorry it seems so hard for me because it has `Animated` feature.

## Checklist

Before you create this PR confirms that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide](https://github.com/dooboolab/hackatalk/blob/master/CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] Run `yarn lint && yarn tsc`
- [x] Run `yarn test` or `yarn test -u` if you need to update snapshot.
- [x] I am willing to follow-up on review comments in a timely manner.

[wehack2021]-[cloud]

hyochan commented:
This is really cool! I think it'd be better if we can have minimal test cases covered for such features like this.

@0916dhkim Could you guide shortly on writing tests here?

0916dhkim commented on server/src/resolvers/User/subscription.ts:
```diff
@@ -37,7 +37,7 @@ export const userUpdated = subscriptionField('userUpdated', {
       assert(userId, 'Not Authorized!');
 
-      return payload.id === userId;
+      return payload.id !== userId;
```
Good catch 👍

0916dhkim reviewed (commented):
@rarira This looks awesome 💯 
For testing, check out existing tests in `client/src/components/pages/__tests__/SignIn.test.tsx` file. The basic structure of each test case is:
1. Prepare fake graphql responses for Relay (if needed).
2. Render the component of interest with appropriate props.
3. Fire user events (if needed).
4. Assert whether the final UI is what we expected. (contains certain text, label etc.)

Good test for `StatusMessageView` component would be checking if the third line is visible when `opened` props is `true` (and `false`). Also, checking for ellipses can be a good test too.

I left some other minor suggestions below 👇

rarira commented on client/src/components/navigations/MainStackNavigator/ProfileModal.tsx:
```diff
@@ -143,6 +138,14 @@ const ModalContent: FC<ModalContentProps> = ({modalState, hideModal}) => {
     useState<boolean>(false);
 
+  // to animate opacity of the modal box
+  const [modalLayout, setModalLayout] = useState<{
```
Indeed, it relates to modal itself not status message component.
```
       \\ line 296

        onLayout={(event) => {
          const {width, height} = event.nativeEvent.layout;
          setModalLayout({width, height});
        }}>
```

rarira commented on client/src/components/navigations/MainStackNavigator/ProfileModal.tsx:
```diff
@@ -143,6 +138,14 @@ const ModalContent: FC<ModalContentProps> = ({modalState, hideModal}) => {
+    height: number;
+  }>({width: 0, height: 0});
+  const [opened, setOpened] = useState(false);
+  const transitionOpacity = useRef(new Animated.Value(0));
```
Same as above, it relates to `Modal View` itself not statusMessage comoponent

rarira commented on client/src/components/navigations/MainStackNavigator/ProfileModal.tsx:
```diff
@@ -275,177 +280,215 @@ const ModalContent: FC<ModalContentProps> = ({modalState, hideModal}) => {
+          backgroundColor: 'black',
+          opacity: transitionOpacity.current,
+        }}
+        {...(!opened && {pointerEvents: 'box-none'})}
```
I think this is the matter of preferences, is it wrong with spread operator?

0916dhkim replied:
@rarira I thought it was strange to spread single property. Your code is not wrong at all.

rarira commented on client/src/components/uis/StatusMessageView.tsx:
```diff
@@ -0,0 +1,136 @@
+    width: number;
+    height: number;
+  };
+  opened: boolean;
```
I'll choose `isStatusMessageExpanded` as suggested above

rarira commented on client/src/components/uis/StatusMessageView.tsx:
```diff
@@ -0,0 +1,136 @@
+import styled from '@emotion/native';
+
+type StyledTextProps = {
+  opened: Boolean;
```
Same as above

rarira commented on client/src/components/uis/StatusMessageView.tsx:
```diff
@@ -0,0 +1,136 @@
+
+  const [textLayoutWidth, setTextLayoutWidth] = useState(0);
+
+  const transition = useRef(new Animated.Value(140)).current;
```
Indeed, it is the first time I wrote animation codes for myself, lots of online references suggest to use `useRef` regardless of `using ref` because of (I think) re-rendering issue. 
https://blog.codecentric.de/en/2019/07/react-native-animated-with-hooks/#:~:text=a%20different%20solution.-,useRef,-useRef%20returns%20a

https://reactjs.org/docs/hooks-reference.html#useref
also mention use case of `useRef` regardless of `using ref`.

rarira commented on client/src/components/uis/StatusMessageView.tsx:
```diff
@@ -0,0 +1,136 @@
+const StatusMessageView: FC<Props> = (props: Props) => {
+  const [bodyHeight, setBodyHeight] = useState(0);
+
+  const [seeMore, setSeeMore] = useState<{show: Boolean; length: number}>({
+    show: false,
+    length: 0,
+  });
```
It is to display arrow button.. if there seems to be more lines of statusMessage than shown basic state which is `numberOfLines={2}`

```
\\\ line 78
  useEffect(() => {
    const statusMessageLength = statusMessage?.split('\n').length;

    if (textLayoutWidth >= 195 || statusMessageLength > 2)
      setShowArrow({show: true, length: statusMessageLength || 0});
    else setShowArrow({show: false, length: statusMessageLength || 0});
  }, [statusMessage, textLayoutWidth]);
```

rarira commented:
> @rarira This looks awesome 💯
> For testing, check out existing tests in `client/src/components/pages/__tests__/SignIn.test.tsx` file. The basic structure of each test case is:
> 
> 1. Prepare fake graphql responses for Relay (if needed).
> 2. Render the component of interest with appropriate props.
> 3. Fire user events (if needed).
> 4. Assert whether the final UI is what we expected. (contains certain text, label etc.)
> 
> Good test for `StatusMessageView` component would be checking if the third line is visible when `opened` props is `true` (and `false`). Also, checking for ellipses can be a good test too.
> 
> I left some other minor suggestions below 👇

Thanks for your detailed reviews and guide for test code.
I'll give it a shot but I'm afraid that it takes a lot of time.
Thanks.

hyochan commented:
> > @rarira This looks awesome 💯
> > For testing, check out existing tests in `client/src/components/pages/__tests__/SignIn.test.tsx` file. The basic structure of each test case is:
> > 
> > 1. Prepare fake graphql responses for Relay (if needed).
> > 2. Render the component of interest with appropriate props.
> > 3. Fire user events (if needed).
> > 4. Assert whether the final UI is what we expected. (contains certain text, label etc.)
> > 
> > Good test for `StatusMessageView` component would be checking if the third line is visible when `opened` props is `true` (and `false`). Also, checking for ellipses can be a good test too.
> > I left some other minor suggestions below 👇
> 
> @0916dhkim Thanks for your detailed reviews and guide for test code.
> I'll give it a shot but I'm afraid that it takes a lot of time.
> Thanks.

I shouldn't take long~! Doesn't have to be perfect and it would be nice if there are minimal test trials in this `PR` for the improvement of the next stages :) You can always ask for trouble~! Cheers 🍻

rarira commented:
@hyochan finally added new test file which  covers 100% of `StatusMessageView` component and updated `ProfileModal` component test codes for the lines which I added to the component. :)

rarira commented:
I made a PR request in dooboo-ui which add a feature to properly display maxLength prop of the EditTexy component. I think it is necessary because the updated showMessage input has maxLength prop.

https://github.com/dooboolab/dooboo-ui/pull/75

rarira commented:
Updated..

1. Change spread operator to ternary operator for the readability 
2. Fix the bug of not showing arrow-down when `StatusMessageView` is expanded
3. Move 'Friends added ' message of `ProfileModal` component to a child component to solve overlapping issue

This is my last commit of this PR.

hyochan reviewed (approved):
@rarira You've rocked :fist_oncoming:
